pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,256 @@
+	ns := system.Namespace()
+	bkts := bs.Buckets()
+	endpointsInformer := endpointsinformer.Get(ctx)
+	f := &Forwarder{
+		selfIP:          selfIP,
+		logger:          logger,
+		kc:              kc,
+		endpointsLister: endpointsInformer.Lister(),
+		bs:              bs,
+		leaseHolders:    make(map[string]string, len(bkts)),
+	}
+
+	leaseInformer := leaseinformer.Get(ctx)
+	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: f.filterFunc(ns),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    f.leaseUpdated,
+			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return f
+}
+
+func (f *Forwarder) filterFunc(namespace string) func(interface{}) bool {
+	return func(obj interface{}) bool {
+		l := obj.(*coordinationv1.Lease)
+
+		if l.Namespace != namespace {
+			return false
+		}
+
+		if !f.bs.HasBucket(l.Name) {
+			// Not for Autoscaler.
+			return false
+		}
+
+		if l.Spec.HolderIdentity == nil || *l.Spec.HolderIdentity == "" {
+			// No holder yet.
+			return false
+		}
+
+		if h, ok := f.getHolder(l.Name); ok && h == *l.Spec.HolderIdentity {
+			// Already up-to-date.
+			return false
+		}
+
+		return true
+	}
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if l.Spec.HolderIdentity == nil {
+		// This shouldn't happen as we filter it out when queueing.
+		// Add a nil point check for safety.
+		f.logger.Warnf("Lease %s/%s with nil HolderIdentity got enqueued", ns, n)
+		return
+	}
+
+	holder := *l.Spec.HolderIdentity
+	f.setHolder(n, holder)
+
+	if holder != f.selfIP {
+		// Skip creating/updating Service and Endpoints if not the leader.
+		return
+	}
+
+	// TODO(yanweiguo): Better handling these errors instead of just logging.
+	if err := f.createService(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Service for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Service for Lease %s/%s", ns, n)
+
+	if err := f.createOrUpdateEndpoints(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Endpoints for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Endpoints for Lease %s/%s", ns, n)
+}
+
+func (f *Forwarder) createService(ns, n string) error {
+	return wait.PollImmediate(retryInterval, retryTimeout, func() (bool, error) {
+		_, err := f.kc.CoreV1().Services(ns).Create(&v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      n,
+				Namespace: ns,
+			},
+			Spec: v1.ServiceSpec{
+				Ports: []v1.ServicePort{{
+					Name:       autoscalerPortName,
+					Port:       autoscalerPort,
+					TargetPort: intstr.FromInt(autoscalerPort),
+				}},
+			},
+		})
+
+		if apierrs.IsAlreadyExists(err) {
+			return true, nil
+		}
+
+		// Do no return the error to make a retry.
+		return err == nil, nil
```

Comment:
might be worth recording the error in a `lastErr` var so we can return it if we fail (rather than just returning a timeout otherwise)

Reply:
done